### PR TITLE
Add edge-weighted PageRank algorithms

### PR DESCRIPTION
**Pull request by mathetake**
I implemented edge-weighted PageRank algorithms in `graph/network` package.

Although there seems to be no related issue, i hope this PR resolve [one of TODOs](https://github.com/gonum/gonum/blob/master/graph/network/network.go#L6).

### CODE REVIEW

File: AUTHORS [hunk 1]
```diff
@@ -61,6 +61,7 @@ source{d} <hello@sourced.tech>
 Shawn Smith <shawnpsmith@gmail.com>
 Spencer Lyon <spencerlyon2@gmail.com>
 Steve McCoy <mccoyst@gmail.com>
+Takeshi Yoneda <cz.rk.t0415y.g@gmail.com>
```
Comment (mathetake) [inline AUTHORS]: If you are welcome, i want to continue to support and contribute to gonum projects.

File: graph/network/page.go [hunk 2]
```diff
@@ -14,6 +14,171 @@ import (
...
+	indexOf := make(map[int64]int, len(nodes))
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+
+	m := mat.NewDense(len(nodes), len(nodes), nil)
+	dangling := damp / float64(len(nodes))
+	for j, u := range nodes {
+		to := g.From(u.ID())
+		z := float64(0)
```
Comment (kortschak) [inline graph/network/page.go]: `var z float64`

File: graph/network/page.go [hunk 3]
```diff
@@ -14,6 +14,171 @@ import (
...
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+
+	m := make(rowCompressedMatrix, len(nodes))
+	var dangling compressedRow
+	df := damp / float64(len(nodes))
+	for j, u := range nodes {
+		to := g.From(u.ID())
+		z := float64(0)
```
Comment (kortschak) [inline graph/network/page.go]: `var z float64`

File: graph/network/page.go [hunk 4]
```diff
@@ -14,6 +14,171 @@ import (
...
+	}
+
+	return ranks
+}
+
+// edgeWeightedPageRankSparse returns the PageRank weights for nodes of the sparse weighted directed
+// graph g using the given damping factor and terminating when the 2-norm of the
+// vector difference between iterations is below tol. The returned map is
+// keyed on the graph node IDs.
+func edgeWeightedPageRankSparse(g graph.WeightedDirected, damp, tol float64) map[int64]float64 {
```
Comment (kortschak) [inline graph/network/page.go]: Why is this not exported?

File: AUTHORS (hunk 1 above)
Comment (kortschak) [reply AUTHORS]: That would be good.

**Comment by kortschak** [comment]
Do you want me to do the type assertion change or will you?

**Comment by mathetake** [comment]
> I think that it would be good for clients to be able to just hand a graph to PageRank or PageRankSparse and have the function type assert on graph.Weighed to use the correct routine.

i fully agree with that, and understand it's hard to do this pr because it involves unexportings.

**Comment by mathetake** [comment]
i did 😄

**Comment by kortschak** [comment]
I'm thinking something more this (on your change).

```diff
diff --git a/graph/network/page.go b/graph/network/page.go
index 1b77850..15f05b7 100644
--- a/graph/network/page.go
+++ b/graph/network/page.go
@@ -14,11 +14,33 @@ import (
        "gonum.org/v1/gonum/mat"
 )
 
-// EdgeWeightedPageRank returns the PageRank weights for nodes of the weighted directed graph g
+// PageRank returns the PageRank weights for nodes of the directed graph g
 // using the given damping factor and terminating when the 2-norm of the
 // vector difference between iterations is below tol. The returned map is
 // keyed on the graph node IDs.
-func EdgeWeightedPageRank(g graph.WeightedDirected, damp, tol float64) map[int64]float64 {
+//
+// If g is a graph.WeightedDirected, an edge-weighted PageRank is calculated.
+func PageRank(g graph.Directed, damp, tol float64) map[int64]float64 {
+       if g, ok := g.(graph.WeightedDirected); ok {
+               return edgeWeightedPageRank(g, damp, tol)
+       }
+       return pageRank(g, damp, tol)
+}
+
+// PageRankSparse returns the PageRank weights for nodes of the sparse directed
+// graph g using the given damping factor and terminating when the 2-norm of the
+// vector difference between iterations is below tol. The returned map is
+// keyed on the graph node IDs.
+//
+// If g is a graph.WeightedDirected, an edge-weighted PageRank is calculated.
+func PageRankSparse(g graph.Directed, damp, tol float64) map[int64]float64 {
+       if g, ok := g.(graph.WeightedDirected); ok {
+               return edgeWeightedPageRankSparse(g, damp, tol)
+       }
+       return pageRankSparse(g, damp, tol)
+}
+
+func edgeWeightedPageRank(g graph.WeightedDirected, damp, tol float64) map[int64]float64 {
        // EdgeWeightedPageRank is implemented according to "How Google Finds Your Needle
        // in the Web's Haystack" with the modification that
        // the columns of hyperlink matrix H are calculated with edge weights.
@@ -101,7 +123,7 @@ func EdgeWeightedPageRank(g graph.WeightedDirected, damp, tol float64) map[int64
 // graph g using the given damping factor and terminating when the 2-norm of the
 // vector difference between iterations is below tol. The returned map is
 // keyed on the graph node IDs.
-func EdgeWeightedPageRankSparse(g graph.WeightedDirected, damp, tol float64) map[int64]float64 {
+func edgeWeightedPageRankSparse(g graph.WeightedDirected, damp, tol float64) map[int64]float64 {
        // edgeWeightedPageRankSparse is implemented according to "How Google Finds Your Needle
        // in the Web's Haystack" with the modification that
        // the columns of hyperlink matrix H are calculated with edge weights.
@@ -179,11 +201,11 @@ func EdgeWeightedPageRankSparse(g graph.WeightedDirected, damp, tol float64) map
        return ranks
 }
 
-// PageRank returns the PageRank weights for nodes of the directed graph g
+// pageRank returns the PageRank weights for nodes of the directed graph g
 // using the given damping factor and terminating when the 2-norm of the
 // vector difference between iterations is below tol. The returned map is
 // keyed on the graph node IDs.
-func PageRank(g graph.Directed, damp, tol float64) map[int64]float64 {
+func pageRank(g graph.Directed, damp, tol float64) map[int64]float64 {
        // PageRank is implemented according to "How Google Finds Your Needle
        // in the Web's Haystack".
        //
@@ -252,11 +274,11 @@ func PageRank(g graph.Directed, damp, tol float64) map[int64]float64 {
        return ranks
 }
 
-// PageRankSparse returns the PageRank weights for nodes of the sparse directed
+// pageRankSparse returns the PageRank weights for nodes of the sparse directed
 // graph g using the given damping factor and terminating when the 2-norm of the
 // vector difference between iterations is below tol. The returned map is
 // keyed on the graph node IDs.
-func PageRankSparse(g graph.Directed, damp, tol float64) map[int64]float64 {
+func pageRankSparse(g graph.Directed, damp, tol float64) map[int64]float64 {
        // PageRankSparse is implemented according to "How Google Finds Your Needle
        // in the Web's Haystack".
        //
diff --git a/graph/network/page_test.go b/graph/network/page_test.go
index 66934b9..350db40 100644
--- a/graph/network/page_test.go
+++ b/graph/network/page_test.go
@@ -91,7 +91,7 @@ func TestPageRank(t *testing.T) {
                                g.SetEdge(simple.Edge{F: simple.Node(u), T: simple.Node(v)})
                        }
                }
-               got := PageRank(g, test.damp, test.tol)
+               got := pageRank(g, test.damp, test.tol)
                prec := 1 - int(math.Log10(test.wantTol))
                for n := range test.g {
                        if !floats.EqualWithinAbsOrRel(got[int64(n)], test.want[int64(n)], test.wantTol, test.wantTol) {
@@ -115,7 +115,7 @@ func TestPageRankSparse(t *testing.T) {
                                g.SetEdge(simple.Edge{F: simple.Node(u), T: simple.Node(v)})
                        }
                }
-               got := PageRankSparse(g, test.damp, test.tol)
+               got := pageRankSparse(g, test.damp, test.tol)
                prec := 1 - int(math.Log10(test.wantTol))
                for n := range test.g {
                        if !floats.EqualWithinAbsOrRel(got[int64(n)], test.want[int64(n)], test.wantTol, test.wantTol) {
@@ -201,7 +201,7 @@ func TestEdgeWeightedPageRank(t *testing.T) {
                                }
                        }
                }
-               got := EdgeWeightedPageRank(g, test.damp, test.tol)
+               got := edgeWeightedPageRank(g, test.damp, test.tol)
                prec := 1 - int(math.Log10(test.wantTol))
                for n := range test.g {
                        if !floats.EqualWithinAbsOrRel(got[int64(n)], test.want[int64(n)], test.wantTol, test.wantTol) {
@@ -232,7 +232,7 @@ func TestEdgeWeightedPageRankSparse(t *testing.T) {
                                }
                        }
                }
-               got := EdgeWeightedPageRankSparse(g, test.damp, test.tol)
+               got := edgeWeightedPageRankSparse(g, test.damp, test.tol)
                prec := 1 - int(math.Log10(test.wantTol))
                for n := range test.g {
                        if !floats.EqualWithinAbsOrRel(got[int64(n)], test.want[int64(n)], test.wantTol, test.wantTol) {
```

**Comment by mathetake** [comment]
truly lgtm

**Comment by mathetake** [comment]
you're, of course, welcome to commit to this PR/branch.

**Comment by kortschak** [comment]
Do you mind applying that change? I'll approve, but we'll wait for someone else as well.

**Comment by mathetake** [comment]
> Do you mind applying that change? I'll approve, but we'll wait for someone else as well.

No I don't mind. I am happy to have your changes.

No, I don't mind.

**Comment by vladimir-ch** [comment]
I will tomorrow.

**Comment by kortschak** [comment]
Thank you @mathetake and welcome.

**Comment by mathetake** [comment]
thank you for your reviews!

**Pull request merged**